ROCm Profiling Database (ROCpd) functionality

adjordje-amd opened the pull request:
This PR implements the rocpd (ROCm Profiling Database) functionality for ROCProfiler-Systems, adding comprehensive database-backed profiling data storage alongside the existing Perfetto and Timemory backends. The implementation introduces a complete database schema, data processing pipeline, and integration points across multiple profiling components.

Key changes include:

- Implementation of rocpd database backend with SQLite storage and comprehensive schema
- Integration of rocpd data collection across sampling, kernel dispatch, memory operations, and system metrics
- Addition of agent management and node information collection for distributed profiling scenarios

adjordje-amd committed:
Add cmake support for sqlite3

adjordje-amd committed:
Add basic sqlite classes

adjordje-amd committed:
Support multiple values in single query

adjordje-amd committed:
Fix ss cleanup

adjordje-amd committed:
make DB singleton

adjordje-amd committed:
Implement data processor wip

adjordje-amd committed:
Use sql statements for writing data

adjordje-amd committed:
Introduce serializable metrics

adjordje-amd committed:
Schema path resolve

adjordje-amd committed:
WIP

adjordje-amd committed:
Work in progress

adjordje-amd committed:
Switch to the new schema - wip

adjordje-amd committed:
move to a new schema

adjordje-amd committed:
Working samples

adjordje-amd committed:
Agent wrapper part1

adjordje-amd committed:
CPU Frequency add tracks

marantic-amd committed:
Record CPU frequency and CPU usage events and samples

marantic-amd committed:
Add overflow and timer event samples to rocpd database. Insert statements for rocpt_regions

marantic-amd committed:
Support for sampling backtrace metrics in rocpd

marantic-amd committed:
Minor PR fixes

Co-authored-by: Aleksandar Djordjevic <aleksandar.djordjevic@amd.com>

marantic-amd committed:
Merge pull request #1 from adjordje-amd/database_sampling_impl

Added database capabilities into overflow and timer sampling

marantic-amd committed:
Store instant kokkos events in rocpd database

adjordje-amd committed:
Generate upid

marantic-amd committed:
MPI/RCCL communication data sizes monitoring

amd-oshkarav committed:
added new schema-views sql files from silopart2-squashed

marantic-amd committed:
Add necessary queries and table entries for tracing kernel dispatch and memcpy

amd-oshkarav committed:
upd data.cpp with the updated schema files

amd-oshkarav committed:
updated data_processor and database.cpp to work with _upid

amd-oshkarav committed:
Update source/lib/core/rocpd/data_storage/database.cpp

Co-authored-by: Aleksandar Djordjevic <aleksandar.djordjevic@amd.com>

amd-oshkarav committed:
Update source/lib/core/rocpd/data_storage/database.cpp

Co-authored-by: Aleksandar Djordjevic <aleksandar.djordjevic@amd.com>

amd-oshkarav committed:
Update source/lib/core/rocpd/data_storage/database.cpp

Co-authored-by: Aleksandar Djordjevic <aleksandar.djordjevic@amd.com>

amd-oshkarav committed:
small changes and tests Alex PR suggestions ok

amd-oshkarav committed:
Merge pull request #2 from adjordje-amd/upd_db_schema

Up to date db schema and code to work with it.

Commit:
Merge remote-tracking branch 'upstream/amd-staging' into amd-staging

amd-oshkarav committed:
removed obsolete tableSchema.sql and utilitySchema.sql

adjordje-amd committed:
Merge remote-tracking branch 'origin/rocprofiler_sdk_rocpd_traces' into amd-staging

adjordje-amd committed:
WIP add set_kernel_rename_and_stream_correlation_id callback

marantic-amd committed:
WiP - Add kernel dispatch and required corresponding tables

marantic-amd committed:
HIP and HSA callback traces in rocpd

marantic-amd committed:
Added Memory Copy traces in rocpd database

adjordje-amd committed:
Rocpd kokkos implementation

Store instant kokkos events in rocpd database

adjordje-amd committed:
Merge pull request #4 from adjordje-amd/rocpd_in_comm_data

MPI/RCCL communication data sizes monitoring

adjordje-amd committed:
Rocprofiler sdk rocpd traces

Rocprofiler sdk rocpd traces

adjordje-amd committed:
Fix view upid

marantic-amd committed:
Initialize table rocpd_metadata and capture callback arguments

adjordje-amd committed:
Fix external data and backtrace

marantic-amd committed:
Add memory allocation traces

marantic-amd committed:
Add memory allocation traces

adjordje-amd committed:
WIP

adjordje-amd committed:
Svasta sam radio

mradosav-amd committed:
WiP Prepare configuration

mradosav-amd committed:
Change database output directory

adjordje-amd committed:
Change database output directory

marantic-amd committed:
Enabled agent_id to be NULL in memory_allocate table. Cleanup

adjordje-amd committed:
Fix missing data

mradosav-amd committed:
Fix flow connections in Perfetto rocpd export (#7)

* Fix missing data

* Fix stack_id, parent_stack_id, corr_id in rocpd

* Remove usage of ancestor for async correlation id

* Fix whitespaces

* Revert

* fix extdata

* add get_parent_stack_id

---------

Co-authored-by: Aleksandar <aleksandar.djordjevic@amd.com>

marantic-amd committed:
Missing category fix

adjordje-amd committed:
Fix code formatting

adjordje-amd committed:
Merge remote-tracking branch 'upstream/amd-staging' into amd-staging

adjordje-amd committed:
Fix agent Smith

mradosav-amd committed:
Align rocpd with new schema (#8)

* Update schema files

* add implementation for get_last_insert_id

* remove region_id, kernel_dispach_id, memory_copy_id from data_processor

* Remove unnecessary IDs from data_processor implementation

* add dependency on use_ram_db for get_last_insert_id

* Fixing thread info handling

* Fix agent handling in database

---------

Co-authored-by: Marjan Antic <marantic@amd.com>

mradosav-amd committed:
fix get_last_insert_row_id for not in-memory db

mradosav-amd committed:
fix stack-id for rocpd-events

mradosav-amd committed:
Fix memory alloc and memory copy values in rocpd

marantic-amd committed:
Improve thread_id and region name id handling in rocprofiler-sdk module (#9)

mradosav-amd committed:
Fix compile errors for MPI

marantic-amd committed:
Agent manager as centralized method for handling GPU and CPU agents (#12)

* Unified usage of agent_manager. Separated adding agents to the manager and to the database

marantic-amd committed:
Use propper logging format

mradosav-amd committed:
Fix crashing of rocprof-sys-avail

mradosav-amd committed:
Add handling of multiple processes for rocpd

mradosav-amd committed:
Add handling of SIGTERM for child process; Include flushing of buffered traces from SDK

mradosav-amd committed:
Reorganize mem_copy domain

marantic-amd committed:
Benchmarking tooling performance

adjordje-amd committed:
Improve benchmark impl

adjordje-amd committed:
Remove bug with insert args

mradosav-amd committed:
Add flush

mradosav-amd committed:
Add handling of multiple processes for rocpd

Add handling of multiple processes for rocpd

marantic-amd committed:
Update sampling module (#16)

adjordje-amd committed:
Add copyright

adjordje-amd commented on .gitignore:
```diff
@@ -42,6 +42,7 @@
 _toc.yml
 
 /build*
+/debug*
```
```suggestion

```

adjordje-amd commented on cmake/SQLite3.cmake:
```diff
@@ -0,0 +1,44 @@
+ else()
+     find_package(SQLite3 REQUIRED)
+     target_link_libraries(rocprofiler-systems-sqlite3 INTERFACE SQLite3::SQLite3)
+ endif()
```
```suggestion
 endif()
 
```

adjordje-amd commented on source/lib/core/benchmark/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+    ${CMAKE_CURRENT_LIST_DIR}/category.hpp)
+
+target_sources(rocprofiler-systems-core-library PRIVATE ${benchmark_sources}
+                                                        ${benchmark_headers})
```
```suggestion
                                                        ${benchmark_headers})

```

adjordje-amd committed:
Apply suggestion from @adjordje-amd

adjordje-amd committed:
Apply suggestion from @adjordje-amd

adjordje-amd committed:
Apply suggestion from @adjordje-amd

mradosav-amd committed:
Add use rocpd flag (#11)

Add flag to enable/disable rocpd

mradosav-amd commented on source/lib/core/categories.hpp:
```diff
@@ -92,10 +92,11 @@ ROCPROFSYS_DEFINE_CATEGORY(category, host, ROCPROFSYS_CATEGORY_HOST, "host", "Ho
 ROCPROFSYS_DEFINE_CATEGORY(category, rocm, ROCPROFSYS_CATEGORY_ROCM, "rocm", "General ROCm tracing")
-ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hip_api, ROCPROFSYS_CATEGORY_ROCM_HIP_API, "rocm_hip_api", "ROCm HIP functions")
-ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hsa_api, ROCPROFSYS_CATEGORY_ROCM_HSA_API, "rocm_hsa_api", "ROCm HSA functions")
+ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hip_api, ROCPROFSYS_CATEGORY_ROCM_HIP_API, "HIP_RUNTIME_API_EXT", "ROCm HIP functions")
```
```suggestion
ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hip_api, ROCPROFSYS_CATEGORY_ROCM_HIP_API, "rocm_hip_api", "ROCm HIP functions")
```

mradosav-amd commented on source/lib/core/categories.hpp:
```diff
@@ -92,10 +92,11 @@ ROCPROFSYS_DEFINE_CATEGORY(category, host, ROCPROFSYS_CATEGORY_HOST, "host", "Ho
-ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hip_api, ROCPROFSYS_CATEGORY_ROCM_HIP_API, "rocm_hip_api", "ROCm HIP functions")
-ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hsa_api, ROCPROFSYS_CATEGORY_ROCM_HSA_API, "rocm_hsa_api", "ROCm HSA functions")
+ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hip_api, ROCPROFSYS_CATEGORY_ROCM_HIP_API, "HIP_RUNTIME_API_EXT", "ROCm HIP functions")
+ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hsa_api, ROCPROFSYS_CATEGORY_ROCM_HSA_API, "HSA_CORE_API", "ROCm HSA functions")
```
```suggestion
ROCPROFSYS_DEFINE_CATEGORY(category, rocm_hsa_api, ROCPROFSYS_CATEGORY_ROCM_HSA_API, "rocm_hsa_api", "ROCm HSA functions")
```

mradosav-amd commented on source/lib/core/categories.hpp:
```diff
@@ -121,7 +122,7 @@ ROCPROFSYS_DEFINE_CATEGORY(category, process_sampling, ROCPROFSYS_CATEGORY_PROCE
 ROCPROFSYS_DEFINE_CATEGORY(category, causal, ROCPROFSYS_CATEGORY_CAUSAL, "causal", "Causal profiling data")
 ROCPROFSYS_DEFINE_CATEGORY(category, cpu_freq, ROCPROFSYS_CATEGORY_CPU_FREQ, "cpu_frequency", "CPU frequency (collected in background thread)")
-ROCPROFSYS_DEFINE_CATEGORY(category, process_page, ROCPROFSYS_CATEGORY_PROCESS_PAGE, "process_page_fault", "Memory page faults in process (collected in background thread)")
+ROCPROFSYS_DEFINE_CATEGORY(category, process_page, ROCPROFSYS_CATEGORY_PROCESS_PAGE, "process_page", "Memory page faults in process (collected in background thread)")
```
Check if name is actually changed

mradosav-amd commented on source/lib/core/rocpd/agent_manager.cpp:
```diff
@@ -0,0 +1,153 @@
+                       type == ROCPROFILER_AGENT_TYPE_GPU ? "GPU" : "CPU");
+
+    std::vector<std::shared_ptr<agent>> agents;
+    for(const auto& agent_ptr : _agents)
```
Change to use `std::copy_if`

mradosav-amd commented on source/lib/rocprof-sys-dl/dl.cpp:
```diff
@@ -147,16 +147,16 @@ reset_rocprofsys_preload()
-        }
-        if(!_modified_preload.empty() && _modified_preload.find(':') == 0)
-            _modified_preload = _modified_preload.substr(1);
+        // auto _modified_preload = std::string{};
```
Should we get rid of this function? Check if there are some side effects for `get_rocprofsys_is_preloaded` and `get_rocprofsys_preload`

mradosav-amd commented on source/lib/rocprof-sys/library.cpp:
```diff
@@ -297,6 +303,73 @@ namespace
+        }
+        ifs.close();
+    }
+
```
```suggestion
    {
        std::string sarg;
        while (std::getline(ifs, sarg, '\0')) {
            _cmdline.push_back(sarg);
        }
        ifs.close();
    }

```

mradosav-amd commented on source/lib/rocprof-sys/library.cpp:
```diff
@@ -297,6 +303,73 @@ namespace
+
+    if(cmd_line.empty())
+    {
+        cmd_line.push_back("rocpd");
```
```suggestion
        cmd_line.push_back("rocprof-systems");
```

mradosav-amd commented on source/lib/rocprof-sys/library.cpp:
```diff
@@ -693,6 +767,15 @@ rocprofsys_finalize_hidden(void)
+            rocprofiler_sdk::shutdown();
+        }
+#endif
+        rocpd::data_processor::get_instance().flush();
```
```suggestion
#if defined(ROCPROFSYS_USE_ROCM) && ROCPROFSYS_USE_ROCM > 0
        // Flush buffered traces in case of child process
        if(get_use_rocm())
        {
            ROCPROFSYS_VERBOSE_F(1, "Shutting down ROCm...\n");
            rocprofiler_sdk::shutdown();

        }
#endif
if(get_use_rocpd()) {
            rocpd::data_processor::get_instance().flush();
            }
```

mradosav-amd commented on source/lib/rocprof-sys/library.cpp:
```diff
@@ -983,6 +1066,7 @@ rocprofsys_finalize_hidden(void)
         [](int) {});
 
     common::destroy_static_objects();
+    rocpd::data_processor::get_instance().flush();
```
```suggestion
   if(get_use_rocpd())
        {
    rocpd::data_processor::get_instance().flush();
    }
```

mradosav-amd commented on source/lib/rocprof-sys/library/amd_smi.cpp:
```diff
@@ -167,9 +302,10 @@ data::sample(uint32_t _dev_id)
+    // TODO: the line below must be uncommented once the amd-smi bug is fixed, otherwise
+    // it causes runtime error on MI30XXX
+    // ROCPROFSYS_AMDSMI_GET(get_settings(m_dev_id).busy, amdsmi_get_gpu_activity,
+    // sample_handle, &m_busy_perc);
```
```suggestion
    ROCPROFSYS_AMDSMI_GET(get_settings(m_dev_id).busy, amdsmi_get_gpu_activity,
    sample_handle, &m_busy_perc);
```

mradosav-amd commented on source/lib/rocprof-sys/library/amd_smi.cpp:
```diff
@@ -69,6 +72,136 @@ using sampler_instances = thread_data<bundle_t, category::amd_smi>;
+        ni.id, getpid(), base_id, TARGET_ARCH, EVENT_CODE, INSTANCE_ID,
+        trait::name<category::amd_smi_memory_usage>::value, "MemUsg",
+        trait::name<category::amd_smi_memory_usage>::description, LONG_DESCRIPTION,
+        COMPONENT, "GB", "ABS", BLOCK, EXPRESSION, 0, 0);
```
```suggestion
        COMPONENT, "MB", "ABS", BLOCK, EXPRESSION, 0, 0);
```

mradosav-amd commented on source/lib/rocprof-sys/library/components/comm_data.cpp:
Include following code snippet in this file
```
void
comm_data::start()
{
    rocpd_initialize_comm_data_categories();
    rocpd_initialize_comm_data_pmc();
}
```

adjordje-amd committed:
Merge remote-tracking branch 'upstream/amd-staging' into amd-staging

mradosav-amd commented on source/lib/rocprof-sys/library/rocprofiler-sdk.cpp:
```diff
@@ -653,7 +1179,7 @@ tool_tracing_callback(rocprofiler_callback_tracing_record_t record,
 #if(ROCPROFILER_VERSION >= 600)
             case ROCPROFILER_CALLBACK_TRACING_OMPT:
-            case ROCPROFILER_CALLBACK_TRACING_MEMORY_ALLOCATION:
+            // case ROCPROFILER_CALLBACK_TRACING_MEMORY_ALLOCATION:
```
Should we bring this back?
```suggestion
            case ROCPROFILER_CALLBACK_TRACING_MEMORY_ALLOCATION:
```

mradosav-amd commented on source/lib/rocprof-sys/library/rocprofiler-sdk.cpp:
```diff
@@ -553,7 +1076,7 @@ tool_tracing_callback(rocprofiler_callback_tracing_record_t record,
 #if(ROCPROFILER_VERSION >= 600)
             case ROCPROFILER_CALLBACK_TRACING_OMPT:
-            case ROCPROFILER_CALLBACK_TRACING_MEMORY_ALLOCATION:
+            // case ROCPROFILER_CALLBACK_TRACING_MEMORY_ALLOCATION:
```
Should we bring this back?
```suggestion
            case ROCPROFILER_CALLBACK_TRACING_MEMORY_ALLOCATION:
```

mradosav-amd commented on source/lib/rocprof-sys/library/rocprofiler-sdk/fwd.hpp:
```diff
@@ -195,6 +209,14 @@ client_data::get_gpu_tool_agent(rocprofiler_agent_id_t id) const
+//         if(gpu_id == itr.device_id) return &itr;
+//     return nullptr;
+// }
+
```
```suggestion
```

mradosav-amd commented on source/lib/core/rocpd/process_info.hpp:
This is not used anywhere

adjordje-amd committed:
Update amd_smi

mradosav-amd committed:
Applied suggestions from code review

mradosav-amd committed:
Include copyright

adjordje-amd committed:
Fix formatting

mradosav-amd committed:
Clean up unused variables & remove process_info class

mradosav-amd committed:
Clean up unused includes

adjordje-amd commented on source/lib/rocprof-sys/library/cpu_freq.cpp:
```diff
@@ -209,47 +429,57 @@ post_process()
         _process_frequencies(_idx, _offset);
     }
-    enabled_cpu_freqs.clear();
+    // enabled_cpu_freqs.clear();
```
Why was this removed?

adjordje-amd committed:
Thread id is optional for track

marantic-amd committed:
Moved cpu_freq track initialization

adjordje-amd committed:
Implement is_optional trait. fix benchmark codestyle

adjordje-amd committed:
Move is_optional implementation

adjordje-amd committed:
Fix typo catergories -> categories

adjordje-amd committed:
Dump all the dbs to the same dir

adjordje-amd committed:
Add rocpd support for
 - cpu_frequency
 - amd_smi
 - sampling

dgaliffiAMD committed:
Adding sqlite-dev to our ci docker images (#280)

sputhala-amd committed:
Make CMakeFile fixes to align with rocDecode and rocJpeg changes (#281)

darren-amd committed:
Fix ROCtx event ranges in trace output (#278)

* Fix marker api traces

* Remove space

* Formatting change

* Small change

* Update Changelog

* Add period to changelog

* Update source/lib/rocprof-sys/library/rocprofiler-sdk.cpp

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

* Fix roctx tests

---------

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

sputhala-amd committed:
Update gotcha submodule from timemory (#277)

* Update gotcha submodule from timemory

* Fix build failure and add copilot suggestions

* Fix formatting errors

adjordje-amd committed:
Fix compile errors, copilot comments

adjordje-amd committed:
Fix shadowed variables

adjordje-amd committed:
Fix agent include when ROCM is not used

dgaliffiAMD committed:
Remove an undefined submodule from the GOTCHA submodule (#285)

A recursive submodule update, `git submodule uppdate --recursive
--init`,
would fail due to an improperly defined and empty submodule in GOTCHA.

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

marantic-amd committed:
Refactor agent manager to work when ROCM is disabled

habajpai-amd committed:
Add missing <cstring> include for C string functions in RCCL tests (#282)

* Fix: Add missing <string.h> include for C string functions in RCCL tests

* Update examples/rccl/rccl-tests/src/common.h

Yes, confirmed—<cstring> alone works in my environment. Updated the PR

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

* clang-format

---------

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

ajanicijamd committed:
Add a CMake Presets file (#44)

Adds the following presets: 

- `ci` - to match the common CI settings - including tests and asserts
- `debug` - True debug build - include building tests
- `debug-optimized` - include building tests
- `release` - To match the "build-release` script - no tests.

The default build folder will be `${sourceDir}/build/<preset>`.

---------

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

marantic-amd committed:
Fix compile errors and typos

marantic-amd committed:
Fix compile errors and typos

adjordje-amd committed:
Move agent and agent_manager from rocpd namespace

adjordje-amd committed:
Merge remote-tracking branch 'refs/remotes/origin/rocpd-implementaion-part-1' into rocpd-implementaion-part-1

adjordje-amd committed:
Replace std::filepath with a timemory filepath impl

adjordje-amd committed:
Fix formatting

adjordje-amd committed:
Fix undefined mpi_send

mradosav-amd committed:
Fix undefined mpi_recv

mradosav-amd committed:
Move use_rocpd guard in comm_data

adjordje-amd committed:
Benchmark support same category from different threads

adjordje-amd committed:
Merge remote-tracking branch 'upstream/rocpd-implementaion-part-1' into amd-staging

adjordje-amd committed:
Cleanup after merge

marantic-amd committed:
Align agent_manager and agents with sdk agent implementations (#20)

adjordje-amd committed:
Merge branch 'amd-staging' into rocpd-implementaion-part-1

adjordje-amd committed:
Code improvement

adjordje-amd committed:
Update source/lib/core/rocpd/data_storage/database.cpp

adjordje-amd committed:
Add ROCPROFSYS_ENABLE_BENCHMARK option

adjordje-amd committed:
Merge remote-tracking branch 'upstream/rocpd-implementaion-part-1' into amd-staging

mradosav-amd committed:
Add cache storage (#17)

Add cache storage

adjordje-amd committed:
Fix compile errors

adjordje-amd committed:
Adjust cache size and threshold

mradosav-amd committed:
Add backward compatability for rocpd post processing

mradosav-amd committed:
Add backward compatability for sdk

mradosav-amd committed:
Add backward compatability for rocprofiler-sdk 6.3.3

mradosav-amd committed:
Cleanup

adjordje-amd committed:
Merge with amd staging (#22)

mradosav-amd committed:
Merge branch 'amd-staging' into backward_compatibility

mradosav-amd committed:
Remove duplicate struct

mradosav-amd committed:
Add backward compatibility

Backward compatibility

mradosav-amd committed:
Add backward compatability for cache (#24)

marantic-amd committed:
Add backward compatability for cache part 2

mradosav-amd committed:
Fix compiler warnings: unused parameters, shadowing etc.

marantic-amd committed:
Fix agent id handling for memory free operations

marantic-amd committed:
Fix agent id handling for memory free operations

ammallya commented:
Error importing PR to ROCm/rocm-systems please reopen the PR there.
